File: Data Structures/Stack/Stack.go
package main

import "fmt"

type Stack []string

// isEmpty: check if stack is empty
func (s *Stack) isEmpty() bool {
	return len(*s) == 0
}

// push a new value onto the stack
func (s *Stack) push(str string) {
	*s = append(*s, str)
}

// Remove and return top element of stack. Return false if stack is empty.
func (s *Stack) pop() (string, bool) {
	if s.isEmpty() {
		return "", false
	} else {
		index := len(*s) - 1
		element := (*s)[index]
		*s = (*s)[:index]
		return element, true
	}
}

func main() {
	var stack Stack // create a stack variable of type Stack

	stack.push("Apple")
	stack.push("Orange")
	stack.push("Grapes")
	stack.push("Watermelons")
	stack.push("Papaya")

	fmt.Println(stack) // Output :- [Apple Orange Grapes Watermelons Papaya]

	stack.pop()
	stack.pop()

	fmt.Println(stack) // Output :- [Apple Orange Grapes]
}

File: Data Structures/Stack/Stack.swift
//
//  Stack.swift
//  
//
//  Created by Pubudu Mihiranga on 19/10/21.
//

import Foundation

struct Stack<T> {
    private var elements: [T] = []
    
    //Add element to stack
    mutating func push(_ element: T) {
        elements.append(element)
    }
    
    //Remove element from stack
    mutating func pop() -> T? {
        return elements.popLast()
    }
    
    //Peeking into the stack is to check the top element of the stack
    func peek() -> T? {
        return elements.last
    }
    
    func printStack() {
        print(elements)
    }
    
}

//MARK: - Usage

var stack = Stack<String>()

stack.push("Lion")
stack.push("Tiger")
stack.push("Whale")
stack.push("Snake")
stack.printStack() //output := ["Lion", "Tiger", "Whale", "Snake"]

let lastElement = stack.peek()
print(lastElement!) //output := Snake

stack.pop()
stack.printStack() //output := ["Lion", "Tiger", "Whale"]

stack.pop()
stack.pop()
stack.pop()
stack.printStack() //output := []
